src/fusiontensor/fusiontensor.jl

```diff
@@ -0,0 +1,189 @@
+matrix_row_axis(ft::FusionTensor) = first(axes(data_matrix(ft)))
+matrix_column_axis(ft::FusionTensor) = last(axes(data_matrix(ft)))
+function charge_block_size(ft::FusionTensor, f1::SectorFusionTree, f2::SectorFusionTree)
+  b = Tuple(findblock(ft, f1, f2))
+  return ntuple(i -> Int(length(axes(ft)[i][b[i]])), ndims(ft))
+end
+
+# GradedUnitRanges interface
+GradedUnitRanges.sector_type(ft::FusionTensor) = sector_type(matrix_row_axis(ft))
+
+# BlockArrays interface
+function BlockArrays.findblock(ft::FusionTensor, f1::SectorFusionTree, f2::SectorFusionTree)
+  # find outer block corresponding to fusion trees
+  @assert ndims_codomain(ft) == length(f1)
+  @assert ndims_domain(ft) == length(f2)
+  @assert sector_type(ft) == sector_type(f1)
+  @assert sector_type(ft) == sector_type(f2)
+  b1 = ntuple(
+    i -> findfirst(==(leaves(f1)[i]), blocklabels(codomain_axes(ft)[i])), ndims_codomain(ft)
+  )
+  b2 = ntuple(
+    i -> findfirst(==(leaves(f2)[i]), blocklabels(dual(domain_axes(ft)[i]))),
+    ndims_domain(ft),
+  )
+  return Block(b1..., b2...)
+end
+
+sanitize_axes(::Tuple{}) = ()
+function sanitize_axes(raw_legs::Tuple{Vararg{AbstractGradedUnitRange}})
+  legs = promote_sectors(typeof(first(raw_legs)), raw_legs)
+  @assert all(check_unique_blocklabels.(legs))
+  return legs
+end
+
+function check_unique_blocklabels(g::AbstractGradedUnitRange)
+  return length(unique(blocklabels(g))) == blocklength(g)
+end
+
+function promote_sectors(
+  ::Type{<:AbstractGradedUnitRange{LA}}, legs::Tuple{Vararg{AbstractGradedUnitRange{LA}}}
+) where {LA}  # nothing to do
+  return legs
+end
+
+function promote_sectors(
+  ::Type{<:AbstractGradedUnitRange}, legs::Tuple{Vararg{AbstractGradedUnitRange}}
+)
+  T = promote_sector_type(legs)
+  # fuse with trivial to insert all missing arguments inside each GradedAxis
+  # avoid depending on SymmetrySectors internals
+  s0 = trivial(T)
+  unified_legs = map_blocklabels.(s -> only(blocklabels(fusion_product(s0, s))), legs)
+  return unified_legs
+end
+
+function promote_sector_type(legs)
+  # fuse trivial sectors to produce unified type
+  # avoid depending on SymmetrySectors internals
+  return sector_type(fusion_product(trivial.(legs)...))
+end
+
+# initialize with already computed data_matrix
+function fusiontensor(
+  mat::AbstractMatrix,
+  codomain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+  domain_legs::Tuple{Vararg{AbstractGradedUnitRange}},
+)
+  # init with empty data_matrix to construct trees_block_mapping
+  ft = FusionTensor(eltype(mat), codomain_legs, domain_legs)
+  @assert space_isequal(matrix_row_axis(ft), axes(mat, 1))
+  @assert space_isequal(matrix_column_axis(ft), axes(mat, 2))
+  for b in eachblockstoredindex(mat)
+    @assert b in eachblockstoredindex(data_matrix(ft))  # check matrix block is allowed
+    data_matrix(ft)[b] = mat[b]
+  end
+  return ft
+end
+
+# empty matrix
+function FusionTensor(
+  data_type::Type,
```

Comment:
I would call this `elt`, i.e. it is the element type (I would interpret `data_type` as `typeof(data_matrix(ft))`).